extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateUserTypeProcessor.java

```diff
@@ -0,0 +1,94 @@
+
+public final class HibernateUserTypeProcessor {
+    private static final String TYPE_VALUE = "type";
+    private static final String TYPE_CLASS_VALUE = "typeClass";
+    private static final DotName TYPE = DotName.createSimple(Type.class.getName());
+    private static final DotName TYPE_DEFINITION = DotName.createSimple(TypeDef.class.getName());
+    private static final DotName TYPE_DEFINITIONS = DotName.createSimple(TypeDefs.class.getName());
+
+    @BuildStep
+    public void build(BuildProducer<ReflectiveClassBuildItem> reflectiveClass, CombinedIndexBuildItem combinedIndexBuildItem) {
+        IndexView index = combinedIndexBuildItem.getIndex();
+
+        final Set<String> userTypes = new HashSet<>();
+
+        Collection<AnnotationInstance> typeAnnotationInstances = index.getAnnotations(TYPE);
+        Collection<AnnotationInstance> typeDefinitionAnnotationInstances = index.getAnnotations(TYPE_DEFINITION);
+        Collection<AnnotationInstance> typeDefinitionsAnnotationInstances = index.getAnnotations(TYPE_DEFINITIONS);
+
+        userTypes.addAll(getUserTypes(typeDefinitionAnnotationInstances));
+
+        for (AnnotationInstance typeDefinitionAnnotationInstance : typeDefinitionsAnnotationInstances) {
+            final AnnotationValue typeDefinitionsAnnotationValue = typeDefinitionAnnotationInstance.value();
+
+            if (typeDefinitionsAnnotationValue == null) {
+                continue;
+            }
+
+            userTypes.addAll(getUserTypes(Arrays.asList(typeDefinitionsAnnotationValue.asNestedArray())));
+        }
+
+        for (AnnotationInstance typeAnnotationInstance : typeAnnotationInstances) {
+            final AnnotationValue typeValue = typeAnnotationInstance.value(TYPE_VALUE);
+            if (typeValue == null) {
+                continue;
+            }
+
+            final String type = typeValue.asString();
+
+            if (!isClass(type)) {
+                continue; // Already registered through TypeDef annotation scanning
+            }
+
+            userTypes.add(type);
+        }
+
+        if (!userTypes.isEmpty()) {
+            reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, userTypes.toArray(new String[] {})));
+        }
+    }
+
+    private Set<String> getUserTypes(Collection<AnnotationInstance> typeDefinitionAnnotationInstances) {
+        final Set<String> userTypes = new HashSet<>();
+
+        for (AnnotationInstance typeDefAnnotationInstance : typeDefinitionAnnotationInstances) {
+            final AnnotationValue typeClassValue = typeDefAnnotationInstance.value(TYPE_CLASS_VALUE);
+            if (typeClassValue == null) {
+                continue;
+            }
+
+            final String typeClass = typeClassValue.asClass().name().toString();
+            userTypes.add(typeClass);
+        }
+
+        return userTypes;
+    }
+
+    private boolean isClass(String type) {
+        try {
+            Class.forName(type);
```

Comment:
I think this should use jandex.

Reply:
@FroMage Thanks. Do you think something like

```
 final String type = typeValue.asString();
 final DotName className = DotName.createSimple(type);
 if (index.getClassByName(className) == null) {
       continue; // Already registered through TypeDef annotation scanning
  }
```  
Will do the trick?

Reply:
I think it should.